modules/nextflow/src/main/groovy/nextflow/extension/PublishOp.groovy

```diff
@@ -105,13 +109,14 @@ class PublishOp {
             final files = entry.value
             publisher.apply(files, sourceDir)
         }
-
         // append record to index file
         if( indexOpts ) {
             final record = indexOpts.mapper != null ? indexOpts.mapper.call(value) : value
             final normalized = normalizePaths(record, targetResolver)
             log.trace "Normalized record for index file: ${normalized}"
             indexRecords << normalized
+            // emit workflow publish event
+            session.notifyWorkflowPublish(name, normalized)
```

Comment:
Something I've been thinking about saving the index metadata to the workflow result. Currently it will always be a list because of how this publish event works. If the source channel is a value channel, it will be saved as a JSON array with a single element rather than just a JSON object

So I would like to refactor this event to instead happen in the onComplete of the publish op, and include either (1) the one record as an object if the source was a value channel or (2) the entire list of index records if the source was a queue channel.

If you think you can implement that, feel free, otherwise I can do it myself

Reply:
I will do it. 
The other doubt I had in this part is about normalized paths. Originally, it was calling notifyWorkflowPublish per value without normalizing the path. It was just normalized for the index file. For testing purposes I moved the notification to this place. But I think it should notify values with normalized paths in all the cases, even the index file is not set. What do you think about it?

Reply:
Yeah I wasn't sure originally whether to normalize the paths here or not, since a trace observer could use the onFilePublish events to recover the source-target mapping.

But as we integrate the index file with the cid store, I think it makes sense to normalize them in the index file.

And yes, I think we should do the onWorkflowPublish regardless of whether the index file is specified. The only complication is the index "mapper", but I plan to remove it anyway in the third preview, so you can remove it here if you want

Reply:
I have updated the [workflow outputs](https://github.com/nextflow-io/nextflow/tree/workflow-outputs-3) branch based on our discussion. So you can just copy the changes from that branch if you want

Reply:
Copied to the branch